### ISSUE TITLE
proxy for outgoing requests

### ISSUE THREAD

**Issue by S2-** [lwsjs/local-web-server#114]
Is there a way to specify a proxy for outgoing requests?
At work we have to pass requests that go to the internet trough a proxy. I have `http_proxy` and `https_proxy` environment variables set, and most command line utilities like `curl`, `git` and so on use those env vars, but when I do something like
```
--rewrite '/api/* -> https://external-service.example.com/api/$1'
```
the `https://external-service.example.com/api/$1` requests do not use the proxy for the outgoing connection.
Am I doing something wrong, or is a proxy just not supported?

**Comment by 75lb** [comment]
Hi, sorry I was on holiday.. requests rewritten to a remote server using `--rewrite` do not currently respect `http_proxy` env variables, I will look into this..

**Comment by S2-** [comment]
No problem @75lb, thanks for the answer and thanks for purring this awesome web server together.
Hope you had a nice holiday!

**Comment by S2-** [comment]
I tried to create a patch for this in [req-then](https://github.com/75lb/req-then), but the proxy connection is rather complex for https endpoint (a `CONNECT` has to be issued first).
Would it be ok for you if I mess around in `req-then` to add the proxy support? Or would you rather use some other library that can handle a proxy to make the outgoing request and add that as a dependency?

**Comment by S2-** [comment]
`local-web-server` could also use [global-tunnel-ng](https://github.com/np-maintain/global-tunnel), just by adding `require('global-tunnel-ng').initialize()` in the `ws` script, or sometime on startup. But the (globalAgent override)[https://github.com/lwsjs/rewrite/blob/master/index.js#L24] in `rewrite` would have to go, and (setSocketKeepAlive)[https://nodejs.org/api/http.html#http_request_setsocketkeepalive_enable_initialdelay] has to be used.

I would be happy to make a pull request. Let me know @75lb.

**Comment by S2-** [comment]
Here is my proposal:
https://github.com/lwsjs/rewrite/pull/2
https://github.com/lwsjs/local-web-server/pull/116

**Comment by 75lb** [comment]
Hi, thanks again for the contribution.. this is high on my list to look at, will get back to you ASAP 👍

**Comment by 75lb** [comment]
hmm, [global-tunnel](https://github.com/np-maintain/global-tunnel) does not support node v11 and above - it recommends using [global-agent](https://github.com/gajus/global-agent) but that doesn't support node v8 (but that's not a blocker).

We need node v12 support at the least. If your patch works using global-agent, let me know. I'm looking for the most minimal solution possible to achieve `http_proxy` support.

Btw, the `require('global-tunnel-ng').initialize()` line you added to local-web-server should instead be added to the top of the lws-rewrite module. Keeps all the changes in one place and avoids tightly coupling local-web-server to its plugins. 👍

**Comment by S2-** [comment]
Here you go: https://github.com/lwsjs/rewrite/pull/3

**Comment by 75lb** [comment]
If you run this script using your own local proxy host and port details does it correctly print the facebook response? 

```js
const http = require('http')

const reqOptions = {
  host: '134.209.178.181', // replace with your proxy host
  port: 3128, // proxy port
  path: 'https://www.facebook.com',
  headers: {
    host: 'www.facebook.com',
    'user-agent': 'nodejs'
  }
}
http.get(reqOptions, res => {
  console.log(res.statusCode, res.headers)
  res.pipe(process.stdout)
})
```

**Comment by S2-** [comment]
No, it does not work. The proxy is configured to use the [CONNECT ](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT) method for TLS connections.

**Comment by 75lb** [comment]
Hi, thanks for that reproduction example - i'll try that in a min.. in the meantime, could you test [this example](https://gist.github.com/75lb/7e00e6c6acebb1e9da85e5b5521a2a26) please. Here are the commands to run the gist: 

```sh
git clone https://gist.github.com/7e00e6c6acebb1e9da85e5b5521a2a26.git tunnel
cd tunnel/
npm install 
node index.js 
```

The script will attempt to load https://www.google.com via your configured `http_proxy` variable (which must be in `http://x.x.x.x:xxxx` format).

**Comment by S2-** [comment]
jup: works.

**Comment by S2-** [comment]
But it does not work with http://www.nossl.net/

**Comment by 75lb** [comment]
yeah, the example only works with HTTPS - for HTTP I would need to use a different module (http-proxy-agent).

I'm looking for a reliable, unintrusive solution which ideally works with node 8,10 and 12, handles Proxy Auth and both HTTP proxying and CONNECT-based tunnelling. Thanks for helping me so far!

Could you try my [example using the request module](https://gist.github.com/75lb/c3a2da31bbec2711074f7fcacf4c6859) please. It will automatically pick up your `http_proxy` env variable.

```sh
git clone https://gist.github.com/c3a2da31bbec2711074f7fcacf4c6859.git request
cd request/
npm install 
node index.js <url to request>
```

If HTTP requests don't work for you, try forcing the request module to always use CONNECT-tunnelling by setting `tunnel: true` in line 4.

**Comment by S2-** [comment]
`request` works, but it's deprecated.

**Comment by 75lb** [comment]
yeah, personally I never use it but in this situation it ticks the most boxes. I'm not using [global-agent](https://www.npmjs.com/package/global-agent), it is massive (includes the 5mb core-js library), seems incomplete (it's less than one month old and a WIP) and is completely unused in the wild.

**Comment by S2-** [comment]
I understand. Then maybe request is the best option. It will work for a lot of years to come for sure.

**Comment by 75lb** [comment]
This is fixed and released in lws-rewrite v1.0.0. The new version uses the latest version of path-to-regexp which has some [breaking changes](https://github.com/pillarjs/path-to-regexp#compatibility-with-express--4x) since the last version.. basically, if your "from" route contains a wildcard, e.g. `/api/*` it should be changed to `/api/(.*)`. Let me know if you have any issues.

To test the latest lws-rewrite, install it locally to your project then run a custom stack e.g. 

```sh
$ npm install lws-rewrite
$ ws --stack ./node_modules/lws-rewrite/ lws-static
```

**Comment by S2-** [comment]
Works like a charm!

**Comment by S2-** [comment]
@75lb, may I ask when you plan to release this new version?

**Comment by 75lb** [comment]
I'm still working on local-web-server v3 but i've published a prerelease, just for you. 

```
$ npm install local-web-server@next
```

Obviously this is still in beta and unstable! Let me know if you have any issues.

**Comment by 75lb** [comment]
Fixed and released in [v3.0.0](https://github.com/lwsjs/local-web-server/releases)
